### PR TITLE
support for triplicate events

### PR DESCRIPTION
**Pull request by mmoskal**
Also adds version bit in frame flags.

### CODE REVIEW

File: inc/interfaces/jd_hw.h [hunk 1]
```diff
@@ -23,6 +23,7 @@ void target_enable_irq(void);
 void target_disable_irq(void);
 void target_wait_us(uint32_t n);
 void target_reset(void);
+int target_in_irq(void);
```
Comment (jamesadevine) [inline inc/interfaces/jd_hw.h:26]: to check for re-entrancy?

File: inc/interfaces/jd_tx.h [hunk 2]
```diff
@@ -11,9 +11,10 @@ int jd_tx_is_idle(void);
 jd_frame_t *jd_tx_get_frame(void);
 void jd_tx_frame_sent(jd_frame_t *frame);
 
-void jd_send(unsigned service_num, unsigned service_cmd, const void *data, unsigned service_size);
-void jd_send_event_ext(srv_t *srv, uint32_t eventid, uint32_t arg);
+int jd_send(unsigned service_num, unsigned service_cmd, const void *data, unsigned service_size);
+void jd_send_event_ext(srv_t *srv, uint32_t eventid, const void *data, uint32_t datalen);
```
Comment (jamesadevine) [inline inc/interfaces/jd_tx.h:15]: is datalen in bytes?

File: inc/interfaces/jd_hw.h (hunk 1 above)
Comment (mmoskal) [reply inc/interfaces/jd_hw.h:26]: yes - the jd_send() apis cannot be called from IRQs

File: source/interfaces/event_queue.c [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    info.qptr -= words << 2;
+    while (words--)
+        *dst++ = *src++;
+}
+
+void jd_process_event_queue(void) {
+    if (info.qptr == 0)
+        return;
+
+    ev_t *ev = info.buffer;
```
Comment (jamesadevine) [inline source/interfaces/event_queue.c:55]: could be dangerous if info.buffer not init'd?

Comment (jamesadevine) [reply source/interfaces/event_queue.c:55]: Is initialisation to zero guaranteed? I guess so if it's in globals...

**Review by jamesadevine — APPROVED** [review]

File: inc/interfaces/jd_tx.h (hunk 2 above)
Comment (jamesadevine) [reply inc/interfaces/jd_tx.h:15]: I think it's in words actually...?

Comment (mmoskal) [reply inc/interfaces/jd_tx.h]: bytes; renamed

File: source/interfaces/event_queue.c (hunk 3 above)
Comment (mmoskal) [reply source/interfaces/event_queue.c:56]: added comment; globals are 0-initalized

**Pull request merged**